### ISSUE TITLE
False negative on nullable analysis

### ISSUE THREAD

**Issue by UniqeId6542** [dotnet/roslyn#82550]
**Version Used**: 
Visual Studio 18.1.1
C# Tools 5.0.0-2.25574.6

**Steps to Reproduce**:
```
internal class DoubleTryGetTest
{
    public static void Test()
    {
        if (TryGet1(out object? o))
        {
            string? s = o.ToString(); // null reference exception produced here.
        }
    }

    private static bool TryGet1([NotNullWhen(true)] out object? o)
    {
        return TryGet2(out o); // no warning
    }

    private static bool TryGet2(out object? o)
    {
        o = null;
        return true;
    }
}
```

**Expected Behavior**:
I expect CS8620 to be made somewhere in this code - likely in the body of TryGet1.

**Actual Behavior**:
No warning is made.

**Comment by RikkiGibson** [comment]
My recollection is that this is by design. See #48425.

Basically we felt that it was "nuisance-prone" to warn in a scenario like impl of `TryGet1`, where we don't know the relationship between the return value and the null state of `o`--versus the case where we do see a relationship (flow state of `o` appears to depend on the return value), and it's an "incorrect" one.

@jcouv in case you have any further thoughts on this.

**Comment by UniqeId6542** [comment]
@RikkiGibson, was the issue that false positive warnings were being made?

If TryGet2 was in an assembly that didn't have nullability enabled, I could see not making CS8620.  But it was in an assembly with it enabled - so I would expect to be warned that the output may be incorrect.

If I add `[NotNullWhen(false)]` to TryGet2, I do get a warning.  So is it the warning ignored only for ambiguous states?  I thought that ambiguous states was a lot of what nullability analysis was for.

**Comment by jcouv** [comment]
There were definitely some compromises we had to make for conditional attributes. Here's an issue that calls out the specific limitation: https://github.com/dotnet/roslyn/issues/44080
I believe that limitation (analysis of non-constant returned bool values) applies to the scenario above.

But from a quick look at the [PR](https://github.com/dotnet/roslyn/pull/41098) for enforcement of conditional attributes within bodies I didn't find a test for this scenario.
So overall I'd say this is a bug (ie. undesirable behavior) but I think we didn't have a solution for it. I'd still want to add an explicit test and confirm this falls into the known gap.
